clicommand/errors.go

```diff
@@ -0,0 +1,73 @@
+package clicommand
+
+import (
+	"errors"
+	"fmt"
+	"os"
+)
+
+// ExitError is used to signal to main.go that the command should exit with a
+// the exit code in `code`.
+type ExitError struct {
+	code  int
+	inner error
+}
+
+func NewExitError(code int, err error) *ExitError {
+	return &ExitError{code: code, inner: err}
+}
+
+func (e *ExitError) Code() int {
+	return e.code
+}
+
+func (e *ExitError) Error() string {
+	return e.inner.Error()
+}
+
+func (e *ExitError) Unwrap() error {
+	return e.inner
+}
+
+func (e *ExitError) Is(target error) bool {
+	terr, ok := target.(*ExitError)
+	return ok && e.code == terr.code
+}
+
+// SilentExitError instructs the main method to not log anything and just exit
+// with status `code`
+type SilentExitError struct {
+	code int
+}
+
+func NewSilentExitError(code int) *SilentExitError {
+	return &SilentExitError{code: code}
+}
+
+// Error prints the exit code, but should not be used in main.go
+func (e *SilentExitError) Error() string {
+	return fmt.Sprintf("silently exited status %d", e.code)
+}
+
+func (e *SilentExitError) Code() int {
+	return e.code
+}
+
+func (e *SilentExitError) Is(target error) bool {
+	terr, ok := target.(*SilentExitError)
+	return ok && e.code == terr.code
+}
+
+func ErrToExitCode(err error) int {
```

Comment:
I'd defensively catch `err == nil` here (and return 0).

Also, if nothing else is named `ExitCode`, that might be a nicer name for this function?

Reply:
So I had a think, and just calling it `ExitCode` is a bit deceptive as it's not a pure function. I've called it `PrintMessageAndReturnExitCode` now. This goes in the wrong direction in terms of method name length, but the right direction for clarity in my opinion.

Reply:
That's fair.